### PR TITLE
Get CSRF token for purge-mta-config

### PR DESCRIPTION
**Pull request by boyan-velinov**
Jira: LMCROSSITXSADEPLOY-861

### CODE REVIEW

File: clients/restclient/rest_client.go [hunk 1]
```diff
@@ -18,13 +22,16 @@ type RestClient struct {
 
 // NewRestClient creates a new Rest client
 func NewRestClient(host, org, space string, rt http.RoundTripper, jar http.CookieJar, tokenFactory baseclient.TokenFactory) RestClientOperations {
-	t := baseclient.NewHTTPTransport(host, getRestURL(org, space), getRestURL(baseclient.EncodeArg(org), baseclient.EncodeArg(space)), rt, jar)
+	t := baseclient.NewHTTPTransport(host, restBaseURL, restBaseURL, rt, jar)
```
Comment (enchobelezirev) [inline clients/restclient/rest_client.go]: Shouldn't this be the same as before? If you change it, the calls to the backend might not contain the org and space parameters or ?

File: clients/restclient/rest_client.go [hunk 2]
```diff
@@ -18,13 +22,16 @@ type RestClient struct {
...
 
-func getRestURL(org, space string) string {
-	return "rest/" + org + "/" + space
+func NewManagementRestClient(host string, rt http.RoundTripper, jar http.CookieJar, tokenFactory baseclient.TokenFactory) RestClientOperations {
+	t := baseclient.NewHTTPTransport(host, csrfRestBaseURL, csrfRestBaseURL, rt, jar)
+	httpRestClient := New(t, strfmt.Default)
+	return &RestClient{baseclient.BaseClient{TokenFactory: tokenFactory}, httpRestClient}
 }
 
 func (c RestClient) PurgeConfiguration(org, space string) error {
```
Comment (enchobelezirev) [inline clients/restclient/rest_client.go]: Could we reuse the method executeRestOperation for the PurgeConfiguration method as well?

File: clients/restclient/rest_client.go [hunk 3]
```diff
@@ -43,3 +50,23 @@ func (c RestClient) PurgeConfiguration(org, space string) error {
 	}
 	return nil
 }
+
+func (c RestClient) getCsrfToken() error {
```
Comment (enchobelezirev) [inline clients/restclient/rest_client.go]: If you want this to be callable from outside the package, you need to declare it with the capital letter: GetCsrfToken :)

File: clients/restclient/rest_client.go [hunk 4]
```diff
@@ -43,3 +50,23 @@ func (c RestClient) PurgeConfiguration(org, space string) error {
...
+	_, err := executeRestOperation(c.TokenFactory, func(token runtime.ClientAuthInfoWriter) (interface{}, error) {
+		return c.Client.Operations.GetCsrfToken(nil, token)
+	})
+	return err
```
Comment (enchobelezirev) [inline clients/restclient/rest_client.go]: Check the err here if is nil. If yes -> return baseclient.NewClientError(err) as in this way the status code and the http response will be contained. If not -> return nil.

File: clients/restclient/rest_client.go [hunk 5]
```diff
@@ -43,3 +50,23 @@ func (c RestClient) PurgeConfiguration(org, space string) error {
...
+}
+
+func (c RestClient) GetSession() error {
+	c.getCsrfToken()
```
Comment (enchobelezirev) [inline clients/restclient/rest_client.go]: The first call returns error as well, shouldn't we check it whether is nil or not?

File: commands/purge_config_command.go [hunk 6]
```diff
@@ -69,3 +80,9 @@ func (c *PurgeConfigCommand) Execute(args []string) ExecutionStatus {
 func (c *PurgeConfigCommand) reportError(err error) {
 	ui.Failed("Could not purge configuration: %v\n", err)
 }
+
+// NewSessionProvider Returns a new SessionProvider - responponsible for giving a unique token each time
+func (c *PurgeConfigCommand) NewSessionProvider(host string) (csrf.SessionProvider, error) {
```
Comment (enchobelezirev) [inline commands/purge_config_command.go]: What about the MtaClient? Will it be problem when you are working with the MtaClient?

File: clients/restclient/rest_client.go (hunk 1 above)
Comment (boyan-velinov) [reply clients/restclient/rest_client.go]: Actually, the RestClient is used only by purge command which has different URL withtout org and space names in it.

File: commands/purge_config_command.go (hunk 6 above)
Comment (boyan-velinov) [reply commands/purge_config_command.go]: Now it works with RestClient. Do you mean to switch it to MtaClient?
I have tried to do so but session is kept for only one client

File: clients/restclient/rest_client.go (hunk 2 above)
Comment (boyan-velinov) [reply clients/restclient/rest_client.go]: Ok

Comment (boyan-velinov) [reply clients/restclient/rest_client.go]: Done

File: clients/restclient/rest_client.go (hunk 4 above)
Comment (boyan-velinov) [reply clients/restclient/rest_client.go]: Done

File: clients/restclient/rest_client.go (hunk 5 above)
Comment (boyan-velinov) [reply clients/restclient/rest_client.go]: Ok but what should we do if it is nil?

File: commands/purge_config_command.go (hunk 6 above)
Comment (enchobelezirev) [reply commands/purge_config_command.go]: I meant: Will it be a problem if I use the method and in the same way I am working with the MtaClient? What about the http calls there? Will they be able to execute POST calls when the MTA Client did not call the /api/v1/csrf-token endpoint?

Comment (boyan-velinov) [reply commands/purge_config_command.go]: MTA Client will continue to call /api/v1/csrf-token so further POST calls are ok

Comment (enchobelezirev) [reply commands/purge_config_command.go]: I understand now. Could we make the GetSession method in the MtaClient to call the endpoint twice(as is now in the REST client)?

**Pull request merged**